bokehjs/src/coffee/core/selection_manager.ts

```diff
@@ -20,39 +20,73 @@ export class SelectionManager extends HasProps {
     })
   }
 
-  source: DataSource
-
-  selector: Selector
-  inspectors: {[key: string]: HasProps}
+  source: ColumnarDataSource
+  inspectors: {[key: string]: Selection}
 
   initialize(): void {
     super.initialize()
-    this.selector = new Selector()
     this.inspectors = {}
   }
 
   select(renderer_views: RendererView[], geometry: Geometry, final: boolean, append: boolean = false): boolean {
-    let did_hit = false
+    // divide renderers into glyph_renderers or graph_renderers
+    const glyph_renderer_views: RendererView[] = []
+    const graph_renderer_views: RendererView[] = []
     for (const r of renderer_views) {
-      did_hit = did_hit || r.hit_test(geometry, final, append)
+      if (r.model.type == 'GlyphRenderer'){
+        glyph_renderer_views.push(r)
+      } else {
+        if (r.model instanceof GraphRenderer){
+          graph_renderer_views.push(r)
+        }
+      }
+    }
+
+    let did_hit = false
+
+    // graph renderer case
+    for (const r of graph_renderer_views) {
+      const hit_test_result = r.model.selection_policy.hit_test(geometry, r)
+      did_hit = did_hit || r.model.selection_policy.do_selection(hit_test_result, r.model, final, append)
     }
+    // glyph renderers
+    if (glyph_renderer_views.length > 0) {
+      const hit_test_result = this.source.selection_policy.hit_test(geometry, renderer_views)
+      did_hit = did_hit || this.source.selection_policy.do_selection(hit_test_result, this.source, final, append)
+    }
```

Comment:
This may be a subject for another PR, but this dichotomy doesn't look good. For example, adding another kind of renderer would be a pain. I suspect that usage of `model.type` is due to recursive imports? This furthermore suggests that this class shouldn't have any knowledge about specific kinds of renderers.

Reply:
Yeah, I think the issue is that selection happens for groups of glyph renderers (that share a data source), rather than for each glyph renderer individually, so the glyph renderer can't be fully responsible for selection like the graph renderer can. There probably is a better solution, but at this point, leaving it for another PR seems ok.